Subject: [PATCH] documented_changed
---
Index: db_actions.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/db_actions.py b/db_actions.py
--- a/db_actions.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/db_actions.py	(date 1777505621026)
@@ -46,6 +46,11 @@
     password = cur.fetchone()['password']
     return password
 
+def get_user_email(user_id):
+    cur.execute("SELECT email FROM users WHERE user_id=%s", [user_id])
+    email = cur.fetchone()['email']
+    return email
+
 #CREATING USER
 def create_user(username, email, password):
     #do-while loop to generate user_id, keep doing until unique one found...
@@ -86,4 +91,33 @@
     #remove all semicolons (;)
     if ";" in input:
         input = input.replace(";", "")
-    return input
\ No newline at end of file
+    return input
+
+
+
+#Database reviews add and get them
+
+#stores which user reviewed what movie
+def add_review(user_id, movie_id, rating, review_text):
+    cur.execute(
+        """ INSERT INTO reviews (user_id, movie_id, rating, review_text) VALUES (%s, %s, %s, %s)"""
+        , [user_id, movie_id, rating, review_text] )
+    conn.commit()
+
+
+#gets the reviews from database
+#joins users and the movie table so the username and titles show instead of ids
+def get_all_reviews():
+    cur.execute(
+        """ SELECT
+            users.username, movies.title, reviews.rating, reviews.review_text, reviews.created_at
+        FROM reviews JOIN users ON  reviews.user_id =  users.user_id JOIN movies ON reviews.movie_id = movies.movie_id
+        ORDER BY  reviews.created_at DESC""" )
+    return cur.fetchall()
+
+
+#dropdown table
+#displays movies in alphabetical order
+def get_movie_options():
+    cur.execute("SELECT movie_id, title FROM movies ORDER BY title")
+    return cur.fetchall()
\ No newline at end of file
Index: page_reviews.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/page_reviews.py b/page_reviews.py
--- a/page_reviews.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/page_reviews.py	(date 1777505771676)
@@ -3,13 +3,113 @@
 from db_actions import *
 
 
+def submit_review(movie_options, movie_choice, rating, review_text, user_id):
+#must be logged in
+    if user_id is None:
+        ui.notify('You must be logged in to review')
+        return
+
+#validation for the review table
+    if movie_choice.value is None:
+        ui.notify('Pick a movie')
+        return
+
+
+
+    if not rating.value:
+        ui.notify('Enter a rating')
+        return
+
+
+
+    try:
+        rating_num = int(rating.value)
+    except ValueError:
+        ui.notify('Rating must be a number')
+        return
+
+
+
+
+    if not 1 <= rating_num <= 5:
+        ui.notify('Rating must be between 1 and 5')
+        return
+
+    try:
+        selected_movie_id = movie_options[movie_choice.value]
+
+        add_review(
+            int(user_id),
+            int(selected_movie_id),
+            rating_num,
+            review_text.value
+        )
+
+        ui.notify('Review added')
+
+#refreshs page to see updated table
+        ui.navigate.to('/reviews')
+
+    except Exception as e:
+        ui.notify(f'Error: {e}')
+
 
 @ui.page('/reviews')
 def reviews():
     uit_banner()
 
-    #TODO, add code that displays ALL reviews (table w/ reviewer's username, reviewed movie, score, and notes)
+#get the logged in user
+    user_id = app.storage.user.get('user_id', None)
+
+#get movies for dropdown table
+    movies = get_movie_options()
+
+
+#creates a list of movies matches the title with id
+    movie_options = { movie['title']: movie['movie_id']
+                      for movie in movies }
+
     with ui.card():
-        pass
+        ui.label('Add Review')
+        if user_id is None:
+            ui.label('You must be logged in to make a review.')
+            ui.button('Go to Login', on_click=lambda: ui.navigate.to('/login'))
+
+#select makes a dropdown table put movie_option into the dropdown
+        else:
+            movie_choice = ui.select( options=list(movie_options.keys()),   label='Movie')
+
+            rating = ui.input('Rating  (1-5)')
+            review_text = ui.textarea('Review')
+
+            ui.button(
+                'Submit Review ',
+                on_click=lambda: submit_review(
+                    movie_options, movie_choice, rating, review_text, user_id
+                )
+            )
+
+#get the reviews
+    try:
+        rows = get_all_reviews()
+    except Exception as e:
+        rows = []
+        ui.notify(f'Could not load reviews: {e}')
+
+#table columns
+    columns = [
+        {'name': 'username', 'label': 'Reviewer', 'field': 'username'},
+        {'name': 'title', 'label': 'Movie', 'field': 'title'},
+        {'name': 'rating', 'label': 'Rating', 'field': 'rating'},
+        {'name': 'review_text', 'label': 'Review', 'field': 'review_text'},
+        {'name': 'created_at', 'label': 'Date', 'field': 'created_at'},
+    ]
+
+
+# display the reviews we got on table
+
+    with ui.card():
+        ui.label('All Reviews')
+        ui.table(columns=columns, rows=rows)
 
     uit_footnote()
\ No newline at end of file
Index: .idea/RateThatMovie2.iml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/RateThatMovie2.iml b/.idea/RateThatMovie2.iml
--- a/.idea/RateThatMovie2.iml	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/.idea/RateThatMovie2.iml	(date 1777498817253)
@@ -4,7 +4,7 @@
     <content url="file://$MODULE_DIR$">
       <excludeFolder url="file://$MODULE_DIR$/.venv" />
     </content>
-    <orderEntry type="jdk" jdkName="Python 3.12 (RateThatMovie2)" jdkType="Python SDK" />
+    <orderEntry type="jdk" jdkName="Python 3.14 (nicegui-lab-starter)" jdkType="Python SDK" />
     <orderEntry type="sourceFolder" forTests="false" />
   </component>
 </module>
\ No newline at end of file
Index: data importer/more importer notes
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/data importer/more importer notes b/data importer/more importer notes
--- a/data importer/more importer notes	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/data importer/more importer notes	(date 1777527156526)
@@ -84,5 +84,30 @@
     UNIQUE (user_id, item_type, item_id)
 )
 
+ALTER TABLE favorites DROP CONSTRAINT favorites_user_id_fkey;
+ALTER TABLE favorites ADD CONSTRAINT favorites_user_id_fkey FOREIGN KEY (user_id) REFERENCES users(user_id) ON DELETE CASCADE;
+
+
+
+
+
+-- Code for watchlists
+CREATE TABLE watchlists(
+	list_id INT,
+	list_name VARCHAR(128),
+	user_id INT REFERENCES users(user_id) ON DELETE CASCADE,
+	creation_date TIME,
+	PRIMARY KEY (list_id)
+);
+CREATE TABLE watchlist_items(
+	list_id INT REFERENCES watchlists(list_id) ON DELETE CASCADE,
+	movie_id INT REFERENCES movies(movie_id) ON DELETE CASCADE,
+	add_date TIMESTAMP,
+	PRIMARY KEY (list_id, movie_id)
+);
+
+
+
+
 
 
Index: page_settings.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/page_settings.py b/page_settings.py
--- a/page_settings.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/page_settings.py	(date 1777500725585)
@@ -2,14 +2,157 @@
 from ui_toolbox import *
 from db_actions import *
 
+#coded w/ assistance from ChatGPT (not much tho)
 
 
 @ui.page('/settings')
 def settings():
     uit_banner()
 
-    #TODO, add code that displays ALL reviews (table w/ reviewer's username, reviewed movie, score, and notes)
+    #Deny access to not logged-in users
+    s = check_login()
+    if s:
+        __settings_body()
+    else:
+        ui.label('You are not logged in.').style('color: red')
+        ui.button("Return Home", on_click=lambda: ui.navigate.to('/'))
+
+    uit_footnote()
+
+
+
+def __settings_body():
+    user_id = app.storage.user.get('user_id', None)
+
     with ui.card():
-        pass
+        ui.label("Settings").style('font-size: 200%')
+        ui.label(f"Your user id is: {user_id}")
+
+        ui.space()
+
+        __section_updateuserinfo(user_id)  #updating username, email, password
+
+        ui.space()
+
+        __section_deleteuser(user_id)       #deleting the user from the site
+
+        ui.space()
+
+        ui.label('More...')
+
+
+
+
+
+
+
+
+
+
+'''
+FUNCTIONS TO UPDATE xxx
+'''
+@ui.refreshable
+def __section_updateuserinfo(user_id):
+    user_username = get_user_username(user_id)
+    user_email = get_user_email(user_id)
+    user_password = get_user_password(user_id)
+
+    ui.label("Change Username, Email, or Password").style('font-size: 140%')
+    with ui.grid(rows=3, columns=3):
+        ui.label(user_username).style('font-size: 100%; border: 1px solid silver; padding: 6px; border-radius: 10px')
+        new_username_box = ui.input('New Username:').props('dense')
+        ui.button('Update Username', on_click=lambda: __try_update_username(new_username_box))
+
+        ui.label(user_email).style('font-size: 100%; border: 1px solid silver; padding: 6px; border-radius: 10px')
+        new_email_box = ui.input('New Email:').props('dense')
+        ui.button('Update Email', on_click=lambda: __try_update_email(new_email_box))
+
+        ui.label(user_password).style('font-size: 100%; border: 1px solid silver; padding: 6px; border-radius: 10px')
+        new_password_box = ui.input('New Password:').props('dense')
+        ui.button('Update Password', on_click=lambda: __try_update_password(new_password_box))
+
+
+# Functions to update username
+def __try_update_username(new_username_box):
+    new_username = security_input(new_username_box.value)   #input validation
+
+    cur.execute("SELECT count(*) FROM users WHERE username=%s", [new_username])
+    s = cur.fetchone()['count']
+
+    #check if username is already shared w/ someone else
+    if s != 0:
+        ui.notify("Please enter a different username", color='negative')
+    else:
+        user_id = app.storage.user.get('user_id', None)
+        cur.execute("UPDATE users SET username=%s WHERE user_id=%s", [new_username, user_id])
+        conn.commit()
+
+        ui.notify("Username Updated", color='positive')
+        __section_updateuserinfo.refresh(user_id)
+
+
+def __try_update_email(new_email_box):
+    new_email = security_input(new_email_box.value)         #input validation
+
+    cur.execute("SELECT count(*) FROM users WHERE email=%s", [new_email])
+    s = cur.fetchone()['count']
+
+    #check if email is already shared w/ someone else
+    if s != 0:
+        ui.notify("Please enter a different email", color='negative')
+    #check if email does not have the '@'
+    elif '@' not in new_email:
+        ui.notify("Please enter a valid email address", color='negative')
+    else:
+        user_id = app.storage.user.get('user_id', None)
+        cur.execute("UPDATE users SET email=%s WHERE user_id=%s", [new_email, user_id])
+        conn.commit()
+
+        ui.notify("Email Updated", color='positive')
+        __section_updateuserinfo.refresh(user_id)
+
+
+def __try_update_password(new_password_box):
+    new_password = security_input(new_password_box.value)   #input validation
+
+    user_id = app.storage.user.get('user_id', None)
+    cur.execute("UPDATE users SET password=%s WHERE user_id=%s", [new_password, user_id])
+    conn.commit()
+
+    ui.notify("Password Updated", color='positive')
+    __section_updateuserinfo.refresh(user_id)
+
+
+'''
+FUNCTIONS TO DELETE USER
+'''
+def __section_deleteuser(user_id):
+    def delete_user():
+        #first, log the user out
+        app.storage.user.pop('user_id')
+        app.storage.user.pop('username')
+
+        #delete user data
+        cur.execute("DELETE FROM users WHERE user_id=%s", [user_id])
+        conn.commit()
+
+        #redirect to home
+        ui.notify("User Deleted", color='negative')
+        ui.navigate.to('/')
+
+    with ui.dialog() as d, ui.card():
+        ui.label("Are you sure you want to delete your account?")
+        with ui.row().classes('w-full justify-between'):
+            ui.button('🗑️Yes', on_click=lambda: delete_user()).props('color=red').style('width: 7vw')
+            ui.space()
+            ui.button('No', on_click=d.close).style('width: 7vw')
+
+    ui.label("Delete User Account").style('font-size: 140%')
+    ui.button('🗑️ Delete', on_click=d.open).props('color=red')
+
+
+
+
+
 
-    uit_footnote()
\ No newline at end of file
Index: .idea/misc.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/misc.xml b/.idea/misc.xml
--- a/.idea/misc.xml	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/.idea/misc.xml	(date 1777498817268)
@@ -3,5 +3,5 @@
   <component name="Black">
     <option name="sdkName" value="Python 3.12 (RateThatMovie2)" />
   </component>
-  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.12 (RateThatMovie2)" project-jdk-type="Python SDK" />
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.14 (nicegui-lab-starter)" project-jdk-type="Python SDK" />
 </project>
\ No newline at end of file
Index: page_homepage.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/page_homepage.py b/page_homepage.py
--- a/page_homepage.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/page_homepage.py	(date 1777500138920)
@@ -16,12 +16,12 @@
             ui.label("Need to express your love or hatred of any movie?")
             ui.label("Well, this website is for individuals like you.")
 
-    # ⭐ Always show dashboard buttons
-    homepage_dashboard()
-
     # Optional login message
     check_login_msg()
 
+    # ⭐ Always show dashboard buttons
+    homepage_dashboard()
+
     uit_footnote()
 
 def homepage_dashboard():
@@ -30,3 +30,15 @@
         ui.button('See Reviews', on_click=lambda: ui.navigate.to('/reviews'))
         ui.button('Search Movies, Actors, and More!', on_click=lambda: ui.navigate.to('/search'))
         ui.button("My Favorites", on_click=lambda: ui.navigate.to('/favorites'))
+        ui.button('See Your Watchlists', on_click=lambda: ui.navigate.to('/watchlists'))
+
+
+
+
+def check_login_msg():
+    s = check_login()
+    username = app.storage.user.get('username', None)  # default if not logged in is None
+    if s:  # true
+        ui.label("Welcome, " + username + ".")
+    else:
+        ui.label("You are not logged in.")
Index: page_watchlists.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/page_watchlists.py b/page_watchlists.py
new file mode 100644
--- /dev/null	(date 1777524156155)
+++ b/page_watchlists.py	(date 1777524156155)
@@ -0,0 +1,351 @@
+from tlbx_imports import *
+from ui_toolbox import *
+from db_actions import *
+
+#coded w/ assistance from ChatGPT (not much tho)
+
+@ui.page('/watchlists')
+def watchlist():
+    uit_banner()
+
+    # Deny access to not logged-in users
+    s = check_login()
+    if s:
+        __watchlists_body()
+    else:
+        ui.label('You are not logged in.').style('color: red')
+        ui.button("Return Home", on_click=lambda: ui.navigate.to('/'))
+
+    uit_footnote()
+
+
+def __watchlists_body():
+    with ui.card() as view_opts:
+        ui.label("Watchlists").style('font-size: 200%')
+        ui.label("This is where users can assemble lists of movies they want to watch.")
+
+        ui.space()
+
+        ui.label("What do you want to do?")
+        with ui.row():
+            ui.button('Modify Lists', on_click=lambda: ui.navigate.to('/watchlists_modify'))
+            ui.space()
+            ui.button('Create List', on_click=lambda: ui.navigate.to('/watchlists_create'))
+            ui.space()
+            ui.button('Delete List', on_click=lambda: ui.navigate.to('/watchlists_delete')).props('color=red')
+
+    ui.space()
+
+    with ui.card() as view_card:
+        ui.label("Here are your watchlists...")
+        __watchlists_view()
+
+# show all watchlists the user has made...
+def __watchlists_view():
+    user_id = app.storage.user.get('user_id', None)
+
+    cur.execute("SELECT * FROM watchlists WHERE user_id=%s", [user_id])
+    watchlists = cur.fetchall()  # all watchlists
+
+    for watchlist in watchlists:
+        #Create a watchlist card for each watchlist, each containing its items
+        with ui.card():
+            wl_id = watchlist['list_id']
+            wl_name = watchlist['list_name']
+
+            cur.execute("SELECT * FROM watchlist_items as a JOIN movies as b " +
+                        "ON a.movie_id = b.movie_id " +
+                        "WHERE list_id=%s ORDER BY add_date DESC", [wl_id])
+            wl_items = cur.fetchall()
+
+            c = [{'name': 'title', 'field': 'title', 'label': "title"},
+                 {'name': 'release_date', 'field': 'release_date', 'label': "release_date"},
+                 {'name': 'runtime', 'field': 'runtime', 'label': "runtime"},
+                 {'name': 'genres', 'field': 'genres', 'label': "genres"},
+                 {'name': 'adult', 'field': 'adult', 'label': "adult"},
+                 {'name': 'budget', 'field': 'budget', 'label': "budget"},
+                 {'name': 'revenue', 'field': 'revenue', 'label': "revenue"},
+                 {'name': 'movie_id', 'field': 'movie_id', 'label': "movie_id"},
+                 {'name': 'add_date', 'field': 'add_date', 'label': "add_date"}]
+
+            ui.label(f"Watchlist: {wl_name}, id: {wl_id}").style('font-size: 140%')
+            ui.table(columns=c, rows=wl_items).props('flat bordered').style('width: 70vw')
+
+    #Something to show if you have no watchlists.
+    if len(watchlists) == 0:
+        ui.label('You have not made any watchlists.').style('color: red')
+
+
+
+
+
+
+@ui.page('/watchlists_create')
+def __watchlists_create():
+    uit_banner()
+
+    # Deny access to not logged-in users
+    s = check_login()
+    if s:
+        __watchlists_create_body()
+    else:
+        ui.label('You are not logged in.').style('color: red')
+        ui.button("Return Home", on_click=lambda: ui.navigate.to('/'))
+
+    uit_footnote()
+
+def __watchlists_create_body():
+    def __try_wl_create():
+        wlc = security_input(wlc_box.value)
+        #check if wlc
+
+        cur.execute("SELECT count(*) FROM watchlists WHERE list_name=%s AND user_id=%s", [wlc, user_id])
+        s = cur.fetchone()['count']
+        if s != 0:
+            ui.notify('Watchlist already exists', color='negative')
+        else:
+            #1st, generate a new id
+            unique_id_found = False
+            nwl_id = 0
+            while(not unique_id_found):
+                nwl_id = random.randint(1, 1000000)
+                cur.execute("SELECT count(*) FROM watchlists WHERE list_id=%s", [nwl_id])
+                matches = cur.fetchone()['count'] #get dictionary, then take int out of dictionary type
+                #if you've found a user_id number that no account uses, then continue on...
+                if matches == 0:
+                    unique_id_found = True
+
+
+            cur.execute("INSERT INTO watchlists (list_id, list_name, user_id, creation_date) "+
+                        "VALUES (%s, %s, %s, (SELECT CURRENT_TIMESTAMP))",
+                        [nwl_id, wlc, user_id])
+            conn.commit()
+            part1.set_visibility(False)
+            part2.set_visibility(True)
+
+            ui.notify('Watchlist created.', color='positive')
+
+
+    user_id = app.storage.user.get('user_id', None)
+
+    with ui.card() as part1:
+        ui.label("Create Watchlist").style('font-size: 200%')
+
+        with ui.row():
+            wlc_box = ui.input("Enter Name")
+            ui.button('Create!', on_click=lambda: __try_wl_create())
+
+    with ui.card() as part2:
+        ui.label("You've created a watchlist")
+        ui.button("Go Home", on_click=lambda: ui.navigate.to('/'))
+
+    part1.set_visibility(True)
+    part2.set_visibility(False)
+
+
+
+@ui.page('/watchlists_modify')
+def __watchlists_modify():
+    uit_banner()
+
+    # Deny access to not logged-in users
+    s = check_login()
+    if s:
+        __watchlists_modify_body()
+    else:
+        ui.label('You are not logged in.').style('color: red')
+        ui.button("Return Home", on_click=lambda: ui.navigate.to('/'))
+
+    uit_footnote()
+
+
+
+def __watchlists_modify_body():
+    #common variables to use
+    user_id = app.storage.user.get('user_id', None)
+    selected_wl = None
+    selected_op = None
+    selected_itm = None
+
+
+    #cards (what get shown)
+    with ui.card() as part_intro:
+        ui.label("What watchlist would you want to modify?")
+
+        c = [{'name': 'list_name', 'field': 'list_name', 'label': "list_name"},
+             {'name': 'list_id', 'field': 'list_id', 'label': "list_id"},
+             {'name': 'creation_date', 'field': 'creation_date', 'label': 'creation_date'}]
+        cur.execute("SELECT list_id, list_name, creation_date FROM watchlists WHERE user_id=%s", [user_id])
+        r = cur.fetchall()
+        part_i_table = ui.table(columns=c, rows=r, selection='single', row_key='list_id', on_select=lambda e:__select_watchlist(e))
+
+        ui.label("What operation would you want to do?")
+        with ui.row():
+            ui.button("Add New Item", on_click=lambda: __select_operation('a'))
+            ui.button("Remove Item", on_click=lambda: __select_operation('r'))
+        with ui.row().classes('w-full justify-end'):
+            ui.button("Next", on_click=lambda: __leave_step_intro())
+
+    with ui.card() as part_add:
+        ui.label("Add new watchlist")
+        c = [{'name': 'title', 'field': 'title', 'label': "title"},
+             {'name': 'release_date', 'field': 'release_date', 'label': "release_date"},
+             {'name': 'runtime', 'field': 'runtime', 'label': "runtime"},
+             {'name': 'genres', 'field': 'genres', 'label': "genres"},
+             {'name': 'adult', 'field': 'adult', 'label': "adult"},
+             {'name': 'budget', 'field': 'budget', 'label': "budget"},
+             {'name': 'revenue', 'field': 'revenue', 'label': "revenue"},
+             {'name': 'movie_id', 'field': 'movie_id', 'label': "movie_id"}]
+        part_a_table = ui.table(columns=c, rows=[], selection='single', row_key='movie_id', on_select=lambda e:__select_item(e), pagination=100).style('width: 73vw')
+        ui.input("Search Table...").bind_value(part_a_table, 'filter')
+        ui.button('Add to List', on_click=lambda: __leave_step_add())
+
+    with ui.card() as part_remove:
+        ui.label("Remove watchlist")
+        c = [{'name': 'title', 'field': 'title', 'label': "title"},
+             {'name': 'release_date', 'field': 'release_date', 'label': "release_date"},
+             {'name': 'runtime', 'field': 'runtime', 'label': "runtime"},
+             {'name': 'genres', 'field': 'genres', 'label': "genres"},
+             {'name': 'adult', 'field': 'adult', 'label': "adult"},
+             {'name': 'budget', 'field': 'budget', 'label': "budget"},
+             {'name': 'revenue', 'field': 'revenue', 'label': "revenue"},
+             {'name': 'movie_id', 'field': 'movie_id', 'label': "movie_id"},
+             {'name': 'add_date', 'field': 'add_date', 'label': "add_date"}]
+        part_r_table = ui.table(columns=c, rows=[], selection='single', row_key='movie_id', on_select=lambda e:__select_item(e))
+        ui.button('Remove', on_click=lambda: __leave_step_remove())
+
+    with ui.card() as part_end:
+        ui.label("Watchlist Modified")
+        ui.button("Go Back to Watchlists", on_click=lambda: ui.navigate.to("/watchlists"))
+
+
+
+    #SELECTIONS, FOR STORING INFORMATION FOR USE
+    def __select_watchlist(e):
+        if len(e.selection) != 0:   #error check
+            nonlocal selected_wl    #global vars don't work here
+            selected_wl = e.selection[0]['list_id']
+            #print(selected_wl)
+    def __select_operation(i):
+        nonlocal selected_op
+        selected_op = i     #'a' = add, 'r' = remove
+    def __select_item(e):
+        if len(e.selection) != 0:   #error check
+            nonlocal selected_itm
+            selected_itm = e.selection[0]['movie_id']
+            #print(selected_itm)
+
+    #PROCESSES, FOR GOING BETWEEN STEPS
+    def __leave_step_intro():
+        #print("globals, ", selected_wl, selected_op)
+        if selected_op is None or selected_wl is None:
+            ui.notify("Fill out all entries", color='negative')
+            return #go back
+
+        #else, enter the next step, based on what was chosen
+        part_intro.set_visibility(False)
+
+        if selected_op == 'a':      #add
+            cur.execute("SELECT * FROM movies" +
+                        " WHERE movie_id NOT IN (SELECT movie_id FROM watchlist_items WHERE list_id=%s)" +
+                        " ORDER BY title ASC", [selected_wl])
+            part_a_table.add_rows(cur.fetchall())
+            part_a_table.update()
+            part_add.set_visibility(True)
+            return
+        elif selected_op == 'r':    #remove
+            cur.execute("SELECT * FROM watchlist_items as a JOIN movies as b ON a.movie_id=b.movie_id WHERE list_id=%s ORDER BY add_date DESC", [selected_wl])
+            part_r_table.add_rows(cur.fetchall())
+            part_r_table.update()
+            part_remove.set_visibility(True)
+            return
+
+    def __leave_step_add():
+        cur.execute("INSERT INTO watchlist_items (list_id, movie_id, add_date) VALUES (%s, %s, (SELECT CURRENT_TIMESTAMP))", [selected_wl, selected_itm])
+        conn.commit()
+        ui.notify("Added item to watchlist", color='positive')
+        part_add.set_visibility(False)
+        part_end.set_visibility(True)
+
+
+
+    def __leave_step_remove():
+        cur.execute("DELETE FROM watchlist_items WHERE list_id=%s AND movie_id=%s", [selected_wl, selected_itm])
+        conn.commit()
+        ui.notify("Deleted item from watchlist", color='positive')
+        part_remove.set_visibility(False)
+        part_end.set_visibility(True)
+
+
+
+    part_intro.set_visibility(True)
+    part_add.set_visibility(False)
+    part_remove.set_visibility(False)
+    part_end.set_visibility(False)
+
+
+
+@ui.page('/watchlists_delete')
+def __watchlists_delete():
+    uit_banner()
+
+    # Deny access to not logged-in users
+    s = check_login()
+    if s:
+        __watchlists_delete_body()
+    else:
+        ui.label('You are not logged in.').style('color: red')
+        ui.button("Return Home", on_click=lambda: ui.navigate.to('/'))
+
+    uit_footnote()
+
+@ui.refreshable
+def __watchlists_delete_body():
+    user_id = app.storage.user.get('user_id', None)
+    wl_ids_names = __get_watchlist_ids_names(user_id)  #get dictionary of key-value pairs (id to name)
+
+    ui.label("Enter list to delete:")
+    selected_wl = ui.select(wl_ids_names)
+
+    with ui.dialog() as d, ui.card():
+        ui.label("Are you sure you want to delete this watchlist?")
+        with ui.row().classes('w-full justify-between'):
+            ui.button('🗑️Yes', on_click=lambda: __try_wl_delete(selected_wl.value, user_id)).props('color=red').style('width: 7vw')
+            ui.space()
+            ui.button('No', on_click=d.close).style('width: 7vw')
+
+    ui.button('Delete List', on_click=d.open).props('color: red')
+
+def __try_wl_delete(s_wl, user_id):
+    #stop if user hadn't put in anything yet
+    if s_wl is None :
+        ui.notify('Enter a list to delete', color='negative')
+        return
+
+    cur.execute("DELETE FROM watchlists WHERE list_id=%s AND user_id=%s", [s_wl, user_id])
+    conn.commit()
+
+    __watchlists_delete_body.refresh()
+    ui.notify('Watchlist deleted', color='positive')
+
+
+
+''' 
+Misc Functionality
+'''
+# find all watchlists you have made, and format them into a dictionary
+# (good for ui.select, because then u get to tie both the name and id together)
+def __get_watchlist_ids_names(user_id):
+    cur.execute("SELECT list_id, list_name FROM watchlists WHERE user_id=%s ORDER BY creation_date", [user_id])
+    list_ids_names = cur.fetchall()
+
+    wl_output = {}
+
+    for l_dn in list_ids_names:
+        l_i = l_dn['list_id']
+        l_n = l_dn['list_name']
+
+        #add new key-value pair
+        wl_output[l_i] = l_n
+
+    return wl_output
\ No newline at end of file
Index: main.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main.py b/main.py
--- a/main.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/main.py	(date 1777499635200)
@@ -22,6 +22,7 @@
 from page_settings import *
 from page_reviews import *
 from page_favorites import *
+from page_watchlists import *
 
 
 #RUN WEBSITE
Index: ui_toolbox.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ui_toolbox.py b/ui_toolbox.py
--- a/ui_toolbox.py	(revision acdf6df0d44610c15e10fcca7e69593aed1c7fab)
+++ b/ui_toolbox.py	(date 1777500725620)
@@ -24,6 +24,7 @@
         ui.button('Home', on_click=lambda: ui.navigate.to('/'))
         ui.button('Login / Logout', on_click=lambda: ui.navigate.to('/login'))
         ui.button('Create Account', on_click=lambda: ui.navigate.to('/create_account'))
+        ui.button('Settings', on_click=lambda: ui.navigate.to('/settings'))
 
 
 def uit_footnote():
@@ -40,13 +41,6 @@
         return False
 
 
-def check_login_msg():
-    s = check_login()
-    username = app.storage.user.get('username', None)  # default if not logged in is None
-    if s:  # true
-        ui.label("Welcome, " + username + ".")
-    else:
-        ui.label("You are not logged in.")
 
 def get_user_id():
     user_id = app.storage.user.get('user_id', None)
